### ISSUE TITLE
Questions on training and stability

### ISSUE THREAD

**Issue by firatozdemir** [microsoft/aurora#70]
When training Aurora, we noticed a recurring (huge) spike in loss values after roughly around 1.8k steps. While trying to debug this, we tried training Aurora from random initialization on WeatherBench2 and noticed the same spike around the same number of steps. We replicate the learning rate strategy of linear warmup of 1k steps followed by half cosine decay based on the paper. We thought it might have to do with high learning rate since it is rather soon after linear warmup is finished. We also use 32 GPUs, so our setup should be very similar.
We do not seem to get this spike if we reduce the max learning rate to 1e-4 instead of 5e-4. 
Do you remember other details to promote stability in your trainings which you can share?

**Comment by wesselb** [comment]
Hey @firatozdemir! Hmm, we have not seen the loss exploding so early in pretraining from a random initialisation. It sounds like something might be going wrong internally in the model.

What would be helpful is to track the magnitude of the model weights and see if some weights become really large. This could indicate where a possible issue is. Would you be able to do this?

We've noticed that the level aggregation mechanism is sometimes unstable. [There is a flag in the architecture which inserts an additional layer norm to stabilise this.](https://microsoft.github.io/aurora/finetuning.html#exploding-gradients) You might want to try this.

**Comment by firatozdemir** [comment]
Hey @wesselb, thank you for your suggestions!
Thanks, we missed this update on the additional layer norm. It unfortunately did not fix our issue. However, you might be onto something with the tracking of model weight norms. I still want to observe it more thoroughly before commenting on it, as the magnitudes are small, but changing rather strangely (I'm seeing a zigzag behavior). 

I wanted to ask another implementation related question. For sharding model gradients (and I assume also model weights), do you use fully sharded data parallel (FSDP) or DeepSpeed Zero-stage 3? We are currently sharding model layers across all 32 GPUs and perhaps there are some low level instabilities causing this kind of behaviour as well.

**Comment by wesselb** [comment]
> We are looking into model weight norms, I will comment if anything seems to change significantly, but on a brief view, it does not seem to be the case.

Hmm, in our experience an the exploding loss is typically preceded by some model weights running off to infinity, which usually indicates where the issue is. If the loss explodes but all model weights are stable, then something else might be wrong, like a numerical precision issue.

> For sharding model gradients (and I assume also model weights), do you use fully sharded data parallel (FSDP) or DeepSpeed Zero-stage 3?

We used FSDP.

**Comment by firatozdemir** [comment]
Hey @wesselb, thanks a lot for your earlier suggestions!
It took us a while to go through some more detailed checks, but I am unfortunately back with a few more questions; the instability early on into pre-training still persists on our side :/.

Can I ask you to verify a few hyper-parameters you used in Aurora for pretraining, it can save us a lot of compute if you happen to know the answers:
1. In `swin3d.py`, in the definition of `Swin3DTransformerBlock`, `scale_bias` term (for the film layer) defaults to zero and the provided script on this repository does not pass any argument. We are wondering if you indeed keep this as zero during training or do you set it to another number. 
2. In the repository, `Aurora` definition; `drop_rate` & `drop_path` defaults to 0.0. However these values are 0.2 in the paper. Can you verify that you used 0.2 for both parameters for pretraining? 
3. Perhaps it's a long stretch, but can you also confirm that you use default pytorch layer initializations (Kaiming uniform) for the `qkv` and `proj` Linear layers inside swin3d layers?

Thanks a lot in advance for your active support!

**Comment by wesselb** [comment]
Hey @firatozdemir!

1. We keep it to the value in the published model, which would be zero.

2. I believe that `drop_path` was set to `0.2`, but `drop_rate` was kept at zero. I'm not entirely sure about this. Let me check and get back to you.

3. Oof, good question. For pretraining, the model was initialised exactly like the model in the repo here is initialised. The initialisations should match.

Do you know precisely which weights blow up first and cause the instability? This is probably the most important thing to figure out.

In which precision are you training? We used PyTorch AMP with `dtype` set to `bfloat16`. Training in pure `bfloat16` might cause stability issues.

**Comment by CalibrationMe** [comment]
Hi @wesselb, thanks a lot for you reply! 

1. In the reference code of AdaptiveLayerNorm (in aurora/model/film.py), the self.scale_bias seems to be 1 instead of 0: https://github.com/facebookresearch/DiT/blob/ed81ce2229091fd4ecc9a223645f95cf379d582b/models.py#L19C31-L20C1

```
def modulate(x, shift, scale):
    return x * (1 + scale.unsqueeze(1)) + shift.unsqueeze(1)
```
Is there any reason why aurora choose to use 0 instead of 1? Also, we did find that using 1 helps a lot for stable training 

2 & 3. Thanks

We did some debug and find that the AdaptiveLayerNorm layer has weird weights patterns and that's why we want to double-check the self.scale_bias value.

we are training with float32, and enable the AMP for the backbone by setting `autocast=True` in Aurora model setting

**Comment by IncubatorShokuhou** [comment]
> Hey [@wesselb](https://github.com/wesselb), thank you for your suggestions! We are looking into model weight norms, I will comment if anything seems to change significantly, but on a brief view, it does not seem to be the case. Thanks, we missed this update on the additional layer norm. It unfortunately did not fix our issue.
> 
> I wanted to ask another implementation related question. For sharding model gradients (and I assume also model weights), do you use fully sharded data parallel (FSDP) or DeepSpeed Zero-stage 3? We are currently sharding model layers across all 32 GPUs and perhaps there are some low level instabilities causing this kind of behaviour as well.

Hi @firatozdemir , I’m trying to fine-tune Aurora on a bunch of 16 GB GPUs (unfortunately, that’s all I have—16 GB each, but I have plenty of them). I noticed you mentioned ZeRO-3 in the issue; could you confirm whether you’re actually using ZeRO-3? If so, could you share a quick demo or example?

**Comment by firatozdemir** [comment]
Hi @IncubatorShokuhou. We use Lightning.Trainer framework. We tried DeepSpeed in the past, but it turned out slower and more problematic than FSDP; it also seems to have less support on lightning.Trainer. So I would recommend FSDP strategy if you are also using lightning.Trainer. 
Finetuning 1.3B Aurora on 16GB GPUs might require a lot of GPUs before it is even possible to fit 1440x720 resolution input..
